swatch-core/src/main/java/org/candlepin/subscriptions/db/SubscriptionCapacityViewRepository.java

```diff
@@ -110,8 +122,49 @@ private SearchCriteria searchCriteriaMatchingUomOfSockets() {
         .build();
   }
 
+  private List<SearchCriteria> searchCriteriaSubscriptionCategories(
+      Set<SubscriptionReportCategory> subsReportCategories) {
+    if (subsReportCategories.containsAll(List.of(SubscriptionReportCategory.values()))) {
+      // Test to see if the category set contains all possible categories. If it does, then there is
+      // no need to add an extra predicate that performs zero actual filtering
+      return Collections.emptyList();
+    }
+    if (subsReportCategories.contains(SubscriptionReportCategory.NON_HYPERVISOR)) {
+      // Non-hypervisor is defined as subs that do not have any virt capacity
+      return List.of(
+          SearchCriteria.builder()
+              .key(SubscriptionCapacityView_.VIRTUAL_SOCKETS)
+              .operation(SearchOperation.EQUAL)
+              .value(0)
+              .build(),
+          SearchCriteria.builder()
+              .key(SubscriptionCapacityView_.VIRTUAL_CORES)
+              .operation(SearchOperation.EQUAL)
+              .value(0)
+              .build());
+    }
+    if (subsReportCategories.contains(SubscriptionReportCategory.HYPERVISOR)) {
+      return List.of(
+          SearchCriteria.builder()
+              .key(SubscriptionCapacityView_.VIRTUAL_SOCKETS)
+              .operation(SearchOperation.GREATER_THAN)
+              .value(0)
+              .build(),
+          SearchCriteria.builder()
+              .key(SubscriptionCapacityView_.VIRTUAL_CORES)
+              .operation(SearchOperation.GREATER_THAN)
+              .value(0)
+              .build());
```

Comment:
The semantics of this are `and`; instead we want `or`. This is most clearly seen in the generated query (captured from debug logging and reformatted):

```
http :8000/api/rhsm-subscriptions/v1/subscriptions/products/RHEL x-rh-identity:$(echo -n '{"identity":{
    "type":"User",
    "user":{"is_org_admin":true},
    "internal":{"org_id":"org123"}
  }}' | base64 -w0) granularity==DAILY beginning==2022-10-01T00:00Z ending==2022-10-05T00:00Z category==hypervisor
```

```sql
select subscripti0_.owner_id            as owner_id1_0_,
       -- more columns, removed for brevity
       subscripti0_.virtual_sockets     as virtual17_0_
from (SELECT sc.subscription_id,
             -- more columns, removed for brevity
             o.description as product_name
      FROM subscription_capacity sc
               JOIN subscription s on sc.subscription_id = s.subscription_id
          AND s.end_date > CURRENT_TIMESTAMP
               JOIN offering o on sc.sku = o.sku) subscripti0_
where subscripti0_.owner_id=?
  and subscripti0_.product_id=?
  and subscripti0_.virtual_sockets > 0
  and subscripti0_.virtual_cores > 0
  and subscripti0_.begin_date<=?
  and subscripti0_.end_date>=?
```

Because our generic searches don't have boolean capabilities (and I think it'd be overkill), it'll probably be necessary to push this into `org.candlepin.subscriptions.db.SubscriptionCapacityViewSpecification#mapSingleSearchCriteriaToPredicate` (maybe it's cleaner to just add a category field to that class?)

Reply:
> Because our generic searches don't have boolean capabilities (and I think it'd be overkill), it'll probably be necessary to push this into `org.candlepin.subscriptions.db.SubscriptionCapacityViewSpecification#mapSingleSearchCriteriaToPredicate` (maybe it's cleaner to just add a category field to that class?)

They do now! :grin: As of #1575